libyul/optimiser/ReasoningBasedSimplifier.h

```diff
@@ -0,0 +1,87 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/Dialect.h>
+
+// because of instruction
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <map>
+
+namespace solidity::smtutil
+{
+class SolverInterface;
+class Expression;
+struct Sort;
+}
+
+namespace solidity::yul
+{
+
+/**
+ * Reasoning-based simplifier.
+ *
+ */
+class ReasoningBasedSimplifier: public ASTModifier
+{
+public:
+	static constexpr char const* name{"ReasoningBasedSimplifier"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	using ASTModifier::operator();
+	void operator()(VariableDeclaration& _varDecl) override;
+	void operator()(If& _if) override;
+
+private:
+	explicit ReasoningBasedSimplifier(
+		Dialect const& _dialect,
+		std::set<YulString> const& _ssaVariables
+	);
+
+	smtutil::Expression encodeExpression(
+		Expression const& _expression
+	);
+
+	smtutil::Expression encodeBuiltin(
+		evmasm::Instruction _instruction,
+		std::vector<Expression> const& _arguments
+	);
+
+	smtutil::Expression int2bv(smtutil::Expression _arg);
+	smtutil::Expression bv2int(smtutil::Expression _arg);
+
+	smtutil::Expression newVariable();
+	smtutil::Expression newRestrictedVariable();
+	std::string uniqueName();
+
+	std::shared_ptr<smtutil::Sort> defaultSort() const;
+	smtutil::Expression constantValue(size_t _value);
+	smtutil::Expression wrap(smtutil::Expression _value);
+
+	Dialect const& m_dialect;
+	std::set<YulString> const& m_ssaVariables;
+	std::unique_ptr<smtutil::SolverInterface> m_solver;
+	std::map<YulString, smtutil::Expression> m_variables;
+
+	bool m_useInt = true;
```

Comment:
It is weird to have this as an option, which is hardcoded. Did you even test this code with `m_UseInt = false`?

Reply:
I left the option as preparation, but can remove

Reply:
Implemented as virtual functions